README: add coreos-assembler bash function, overhaul docs

dustymabe opened the pull request:
This adds the cosa bash function, which brings support for various
environment variables to tweak functionality at runtime. Having this
functionality allows us to overhaul the rest of the documentation to
use it instead of adding new aliases in various places.

dustymabe commented:
example of executing using the function:

```
$ export COREOS_ASSEMBLER_GIT=/var/b/shared/code/github.com/coreos/coreos-assembler/
$ export COREOS_ASSEMBLER_CONFIG_GIT=/var/b/shared/code/github.com/coreos/fedora-coreos-config/
$ export COREOS_ASSEMBLER_PRIVILEGED=
$ export COREOS_ASSEMBLER_CONTAINER=localhost/coreos-assembler
$ export COREOS_ASSEMBLER_CONTAINER_RUNTIME_ARGS=
$
$ cosa() {
>     env | grep COREOS_ASSEMBLER
>     set -x # so we can see what command gets run
>     sudo podman run --rm -ti -v ${PWD}:/srv/ --userns=host --device /dev/kvm --name cosa            \
>                ${COREOS_ASSEMBLER_PRIVILEGED:+--privileged}                                         \
>                ${COREOS_ASSEMBLER_CONFIG_GIT:+-v $COREOS_ASSEMBLER_CONFIG_GIT:/srv/src/config/:ro}  \
>                ${COREOS_ASSEMBLER_GIT:+-v $COREOS_ASSEMBLER_GIT/src/:/usr/lib/coreos-assembler/:ro} \
>                ${COREOS_ASSEMBLER_CONTAINER_RUNTIME_ARGS}                                           \
>                -e COREOS_ASSEMBLER* \
>                ${COREOS_ASSEMBLER_CONTAINER:-quay.io/coreos-assembler/coreos-assembler:latest} $@
>     set +x
> }
$
$ cosa fetch
COREOS_ASSEMBLER_CONTAINER=localhost/coreos-assembler
COREOS_ASSEMBLER_PRIVILEGED=
COREOS_ASSEMBLER_CONTAINER_RUNTIME_ARGS=
COREOS_ASSEMBLER_GIT=/var/b/shared/code/github.com/coreos/coreos-assembler/
COREOS_ASSEMBLER_CONFIG_GIT=/var/b/shared/code/github.com/coreos/fedora-coreos-config/
+ sudo podman run --rm -ti -v /var/b/shared/fedora-assembler:/srv/ --userns=host --device /dev/kvm --name cosa -v /var/b/shared/code/github.com/coreos/fedora-coreos-config/:/srv/src/config/:ro -v /var/b/shared/code/github.com/coreos/coreos-assembler//src/:/usr/lib/coreos-assembler/:ro -e 'COREOS_ASSEMBLER*' localhost/coreos-assembler fetch
info: Missing CAP_SYS_ADMIN; using virt
Using manifest: /srv/src/config/manifest.yaml
Running: rpm-ostree compose tree --repo=/srv/repo --cachedir=/srv/cache --touch-if-changed /srv/tmp/treecompose.changed --unified-core /srv/src/config/manifest.yaml --download-only
```

jlebon commented:
Isn't this what the `bottlecap` script is supposed to wrap?

cgwalters commented:
This would break the symmetry that today one can invoke e.g. `coreos-assembler build` also from *inside* the container via `coreos-assembler shell`.  Today one can use most of the commands in `README.md` both ways (though I do feel we should be encouraging people to run most commands from inside the container).

Why specifically did you choose `cosa`?  I feel like there has to be a well-understood rationale to have the binary name not match the project name.  At least currently `coreos-<TAB>` auto-completes here.  (But as soon as some other project was introduced that was `coreos-X` and lived in `/usr/bin` the argument for shortening or changing coreos-assembler would be stronger)

cgwalters commented on README.md:
Oh, you did write this here.  I dunno...particularly inside the container it's `coreo<TAB>` looks like.

ashcrow commented:
> Isn't this what the bottlecap script is supposed to wrap?

I actually never checked that file until this comment. It does look like it.

ajeddeloh commented:
@jlebon yeah, although it's somewhat orthogonal; there's nothing that says you should use bottlecap and nothing that says you shouldn't. Really we should have two sections, one with bottlecap and one without. There are some difference with this approach (bottlecap runs `make` && `make check` && `make install` rather than binding over the `/usr/lib/coreos-assembler` directory. This means you'll catch shellcheck errors as you develop and if we added scripts/helpers requiring compilation it would automatically compile them.

cgwalters commented on README.md:
Let's not document creating world-writable directories.  Generally ACLs are better where necessary.  But in this case I think we can solve this better.

dustymabe replied:
added facl commands to allow write access for the builder user.

cgwalters commented:
> change the readme back to coreos-assembler everywhere - i'm cool with this as it was kind of a late addition

I'd vote for that and circle back to the naming/typing issue as a separate PR.  It's a bit ironic because I find rpm-ostree annoying to type and have been advocating a rename partially based on that but you're on the other side there :wink: 

(I'm fine with `cosa` though it does open questions around whether we're renaming the github project too, how far it permeates into the docs...)

dustymabe commented:
> I'd vote for that and circle back to the naming/typing issue as a separate PR.

pushed fixup ^^

dustymabe commented:
> I'd vote for that and circle back to the naming/typing issue as a separate PR.

opened https://github.com/coreos/coreos-assembler/issues/291 to discuss abbreviation. working on the dir permissions change now.

cgwalters commented on README.md:
```diff
@@ -16,39 +16,126 @@ and to keep them as similar as possible:
+that we support running in a privileged or unprivileged mode (detailed
+below). Regardless of whether you are using privileged or unprivileged
+mode you'll need access to `/dev/kvm` so that we can build the resulting
+cloud image using anaconda. If you're running this in a VM, you must enable
```
Minor, but how about:

> you'll need access to `/dev/kvm` as the build process runs a virtual machine in order to generate the target image.

dustymabe commented on README.md:
```diff
@@ -16,90 +16,192 @@ and to keep them as similar as possible:
+note you will need a maipo directory with references to
+proper repositories.
+
+You can upload your built image to a registry such as quay.io by doing the following:
```
went with the "You can optionally upload ..." for now.

rfairley replied:
I think I added the `podman push` section a while back, when first trying it out. I was wanting to try building the image myself, uploading it, and then passing the image as `quay.io/rfairley/coreos-assembler:latest` into the `coreos-assembler` alias. At the time, `podman push` was new to me and I saw it as worth documenting. But local building and just passing the image tag for the locally built image is definitely the way to go - we can remove it to focus the section more on local dev.

dustymabe replied:
Thanks. I removed the `podman push` section in a 2nd commit.

dustymabe commented:
ok should have addressed all comments now and rebased/squashed everything so it should be ready for merge if everything looks good.

side note: if someone could answer [this question](https://github.com/coreos/coreos-assembler/issues/292) that would be great.

cgwalters commented:
Overall this looks good...I just wanted to get a chance to test it myself and haven't been at a place with good enough Internet to pull the latest cosa container.  I should be able to do so tomorrow; if someone else beats me to it and tests this they should feel free to merge too!

cgwalters commented:
This replaces https://github.com/coreos/coreos-assembler/pull/263 right?

dustymabe commented:
> .I just wanted to get a chance to test it myself and haven't been at a place with good enough Internet to pull the latest cosa container.

👍  - no rush - i want some of the other people who had comments to sign off too



> This replaces #263 right?

Not exactly. I think @jlebon was trying to document how to use rootless podman to run a build. While we technically can run in a rootless container (assuming you provide /dev/kvm) I haven't found the rootless podman builds/runs to work for me reliably yet, so I don't think we should document that flow. All of the commands in this documentation still use `sudo`.

cgwalters commented on README.md:
Is this intended to be something cut-and-paste-able into a running bash shell?  I get syntax errors, I think because `{` has to be on a newline or something?

dustymabe replied:
yes, intended to be cut/pasted into a shell. Did you include the PS1 prompt (`$`) in your copy/paste? I can remove that if it's confusing.

cgwalters replied:
We should indeed drop the `$` and also document that the text can be cut-and-paste into bash.  I kept doing that copy-paste and it took me a surprisingly long time to realize the problem: bash functions can't have `-` in their name...

So...this kind of brings us right back to the rename...

cgwalters replied:
Tactically for now though just change it to `coreos_assembler`?

cgwalters replied:
(There's also a lesson here about rewriting a PR in response to feedback and testing)

dustymabe replied:
>  I kept doing that copy-paste and it took me a surprisingly long time to realize the problem: bash functions can't have `-` in their name

works for me: 

<details>

```
[dustymabe@media ~]$ coreos-assembler() {
>     env | grep COREOS_ASSEMBLER
>     set -x # so we can see what command gets run
>     sudo podman run --rm -ti -v ${PWD}:/srv/ --userns=host --device /dev/kvm --name coreos-assembler \
>                ${COREOS_ASSEMBLER_PRIVILEGED:+--privileged}                                          \
>                ${COREOS_ASSEMBLER_CONFIG_GIT:+-v $COREOS_ASSEMBLER_CONFIG_GIT:/srv/src/config/:ro}   \
>                ${COREOS_ASSEMBLER_GIT:+-v $COREOS_ASSEMBLER_GIT/src/:/usr/lib/coreos-assembler/:ro}  \
>                ${COREOS_ASSEMBLER_CONTAINER_RUNTIME_ARGS}                                            \
>                ${COREOS_ASSEMBLER_CONTAINER:-quay.io/coreos-assembler/coreos-assembler:latest} $@
>     rc=$?; set +x; return $rc
> }
[dustymabe@media ~]$ which coreos-assembler 
coreos-assembler ()
{ 
    env | grep --color=auto COREOS_ASSEMBLER;
    set -x;
    sudo podman run --rm -ti -v ${PWD}:/srv/ --userns=host --device /dev/kvm --name coreos-assembler ${COREOS_ASSEMBLER_PRIVILEGED:+--privileged} ${COREOS_ASSEMBLER_CONFIG_GIT:+-v $COREOS_ASSEMBLER_CONFIG_GIT:/srv/src/config/:ro} ${COREOS_ASSEMBLER_GIT:+-v $COREOS_ASSEMBLER_GIT/src/:/usr/lib/coreos-assembler/:ro} ${COREOS_ASSEMBLER_CONTAINER_RUNTIME_ARGS} ${COREOS_ASSEMBLER_CONTAINER:-quay.io/coreos-assembler/coreos-assembler:latest} $@;
    rc=$?;
    set +x;
    return $rc
}
[dustymabe@media ~]$ mkdir /tmp/test
[dustymabe@media ~]$ cd /tmp/test
[dustymabe@media test]$ coreos-assembler shell
+ sudo podman run --rm -ti -v /tmp/test:/srv/ --userns=host --device /dev/kvm --name coreos-assembler quay.io/coreos-assembler/coreos-assembler:latest shell
bash: cannot set terminal process group (-1): Inappropriate ioctl for device
bash: no job control in this shell
[coreos-assembler]$ 
[coreos-assembler]$ exit
exit
+ rc=0
+ set +x

```


</details>

cgwalters replied:
Ohh...argh, I just figured it out.  I still had my old `coreos-assembler` alias in that shell.

dustymabe replied:
😢 - I'm sorry this wasted some time. Thanks for testing, though

rfairley commented:
Gave this a run, and works well overall! Ran into a couple of problems, which could be specific to my environment.

I needed to set `COREOS_ASSEMBLER_PRIVILEGED=true` for `coreos-assembler run` to work on my end (after doing `init`, `fetch`, and `build` unpriv successfully). The avc denial doing `run` with unprivileged was (see Details):

<details>
```
$ coreos-assembler run
...
+ exec qemu-kvm -name coreos -m 2048 -nographic -netdev user,id=eth0,hostname=coreos -device virtio-net-pci,netdev=eth0 -object rng-random,filename=/dev/urandom,id=rng0 -device virtio-rng-4
qemu-system-x86_64: -drive if=virtio,cache=unsafe,file=builds/29/fedora-coreos-29-qemu.qcow2: Could not open 'builds/29/fedora-coreos-29-qemu.qcow2': Permission denied
+ rc=1
+ set +x

$ sudo ausearch -m avc --start 17:14:55
----
time->Tue Jan 29 17:14:55 2019
type=AVC msg=audit(1548800095.064:1972): avc:  denied  { read } for  pid=13784 comm="qemu-system-x86" name="fedora-coreos-29-qemu.qcow2" dev="dm-0" ino=528736 scontext=system_u:system_r:co0
----
time->Tue Jan 29 17:14:55 2019
type=AVC msg=audit(1548800095.064:1973): avc:  denied  { read } for  pid=13784 comm="qemu-system-x86" name="fedora-coreos-29-qemu.qcow2" dev="dm-0" ino=528736 scontext=system_u:system_r:co0

$ sudo ausearch -m avc --start 17:14:55 | audit2allow


#============= container_t ==============

#!!!! This avc is a constraint violation.  You would need to modify the attributes of either the source or target types to allow this access.
#Constraint rule: 
#	mlsconstrain file { ioctl read lock execute execute_no_trans } ((h1 dom h2 -Fail-)  or (t1 != mcs_constrained_type -Fail-) ); Constraint DENIED
mlsconstrain file { write setattr append unlink link rename } ((h1 dom h2 -Fail-)  or (t1 != mcs_constrained_type -Fail-) ); Constraint DENIED
mlsconstrain file { create relabelto } ((h1 dom h2 -Fail-)  and (l2 eq h2)  or (t1 != mcs_constrained_type -Fail-) ); Constraint DENIED
mlsconstrain file { relabelfrom } ((h1 dom h2 -Fail-)  or (t1 != mcs_constrained_type -Fail-) ); Constraint DENIED

#	Possible cause is the source level (s0:c585,c1005) and target level (s0:c174,c429) are different.
allow container_t container_file_t:file read;

```
</details>

rfairley commented:
One other thing, after `export COREOS_ASSEMBLER_PRIVILEGED=true` and then `coreos-assembler run`, the MOTD (with the QEMU monitor and ICMP messages) didn't show nor was the unit to generate it found:

<details>

```
Fedora 29 (CoreOS preview)
Kernel 4.20.4-200.fc29.x86_64 on an x86_64 (ttyS0)

This is coreos (Linux x86_64 4.20.4-200.fc29.x86_64) 22:51:45
SSH host key: SHA256:gUQbUlB4v1PLnKmKuE7JsjKMZw+YYuygwfyBBEGLcLk (ED25519)
SSH host key: SHA256:3uBQrrfqn8E0cz7bDei6q83uk9CKvrSzzls2kZCt9yY (ECDSA)
SSH host key: SHA256:C4FXCZFel7m4ooXwsVfRxQYudUDPyPEhdCL0VrC1+68 (RSA)
eth0: 10.0.2.15 fec0::5054:ff:fe12:3456

coreos login: core (automatic login)

Fedora CoreOS (preview)
https://github.com/coreos/fedora-coreos-tracker
WARNING: All aspects subject to change, highly experimental
---

[core@coreos ~]$ systemctl status serial-getty-motd
Unit serial-getty-motd.service could not be found.
```

</details>

Not sure if the unit not getting in could be related to running unpriv then priv, or if it's something with my environment - will check on laptop too tomorrow.

rfairley commented on README.md:
typo for "directory"

dustymabe replied:
fixed

dustymabe commented:
> @rfairley 
> The avc denial doing `run` with unprivileged was

I've seen this before and am attempting to fix this in https://github.com/coreos/coreos-assembler/pull/294

> @rfairley 
>  the MOTD (with the QEMU monitor and ICMP messages) didn't show nor was the unit to generate it found

haven't seen that myself - might have to do some investigation after this merges if you still see it.

rfairley commented:
@dustymabe 

Just to follow up on the second issue (MOTD), `coreos-assembler run` worked on my laptop (a new install of Silverblue), so it does seem like just something on the desktop (F29 where some networking config has been changed by me at some point). Will check later to see if there is any log from Ignition indicating a  problem.

cgwalters commented on README.md:
It'd be a bit nicer if we did something like auto-detect the case where this is a symlink and did a the bind mount over it or so.  But not necessary to do in this change.

dustymabe replied:
> It'd be a bit nicer if we did something like auto-detect the case where this is a symlink and did a the bind mount over it or so.

yeah, hadn't thought of that before. now that we're using a function and not an alias we could do that

> But not necessary to do in this change.

is that a +1 for merge?

dustymabe commented:
anything else outstanding for this review?

jlebon commented on README.md:
Minor/optional: I'd just make these `# comment lines` in the previous block to make it easier to follow.

dustymabe commented:
> Just one minor thing I commented above

Thanks. I removed the `podman push` section in a 2nd commit.

dustymabe commented:
> A minor detail, just seeing this now: should the paths `/srv/coreos` be updated to `./srv-coreos`?

fixed this in a new commit, I changed the paths to be relative to the host directory PWD

rfairley commented on README.md:
```diff
@@ -171,8 +171,8 @@ And now we can build from these inputs:
 
-Each build will write an OSTree commit into `/srv/coreos/repo` as well
-as generate VM images in `/srv/coreos/builds/`.
+Each build will write an OSTree commit into `$PWD/src/repo/` as well
```
Should this be `$PWD/repo/`?

dustymabe replied:
fixed

rfairley commented:
:+1: all LGTM, good to merge.

dustymabe commented:
ok going to merge this now since I've got three 👍 . 

I would appreciate comments/votes over in https://github.com/coreos/coreos-assembler/issues/291 regarding the convenience discussion (could be a naming discussion depending on how you look at it).
